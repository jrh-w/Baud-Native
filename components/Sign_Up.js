import { AppLoading } from 'expo';
import { Footer, StyleProvider, Container, Text, Header, Content, Card, CardItem, Body, Thumbnail, H1, Item, Input, Button, Icon } from 'native-base';
import * as Font from 'expo-font';
import { Ionicons, EvilIcons } from '@expo/vector-icons';
import React, { Component } from 'react';
import { Col, Row, Grid } from 'react-native-easy-grid';

import { StatusBar, Image, Dimensions } from 'react-native';
import getTheme from '../native-base-theme/components';
import material from '../native-base-theme/variables/material';

import AppHeader from './sub_components/AppHeader';
import SignUpAlert from './sub_components/SignUpAlert';

import axios from 'axios';
import bcrypt from 'react-native-bcrypt';

class Sign_Up extends Component {
  constructor(props) {
    super(props);
    this.state = {
      //isReady: false
      username: '',
      email: '',
      password: '',
      confirmPassword: '',
      usernameError: 'Username should contain at least 8 characters',
      emailError: 'emailError',
      passwordError: 'passwordError',
      confirmPasswordError: 'confirmPasswordError',
      errorText: '',
      invalidData: false
    };

    this.register = this.register.bind(this);
    this.checkData = this.checkData.bind(this);
  }

  /*async componentDidMount() {
    await Font.loadAsync({
      Montserrat_Regular: require('../node_modules/native-base/Fonts/Foundation.ttf'),
      Montserrat_Bold: require('../assets/fonts/Montserrat/Montserrat_Bold.ttf'),
      SemiBold: require('../assets/fonts/Montserrat/Montserrat-SemiBold.ttf'),
      ...Ionicons.font,
    });
    this.setState({ isReady: true });
  }*/

  checkData(usernameTest, passwordTest, emailTest, passwordsMatch) {
    let checkSuccessful = true;
    this.state.errorText = '';

    switch(false){
      case usernameTest:
        this.state.errorText = this.state.usernameError;
        checkSuccessful = false;
        break;
      case emailTest:
        this.state.errorText = this.state.emailError;
        checkSuccessful = false;
        break;
      case passwordTest:
        this.state.errorText = this.state.passwordError;
        checkSuccessful = false;
        break;
      case passwordsMatch:
        this.state.errorText = this.state.confirmPasswordError;
        checkSuccessful = false;
        break;
      default:
        break;
    }

    /*if(!usernameTest) {
        this.state.errorText = this.state.usernameError;
        checkSuccessful = false;
    }
    else
    if(!passwordTest) {
        this.state.errorText = this.state.passwordError;
        checkSuccessful = false;
    }
    else
    if(!emailTest) {
        this.state.errorText = this.state.emailError;
        checkSuccessful = false;
    }
    else
    if(!passwordsMatch) {
        this.state.errorText = this.state.confirmPasswordError;
        checkSuccessful = false;
    }*/

    this.setState({invalidData : !checkSuccessful})
    console.log('invalidData ' + this.state.invalidData + ' errorText ' + this.state.errorText);
    return checkSuccessful;
  }

  register() {

    let usernameRegEx = new RegExp('^(?=[a-zA-Z0-9._-]{8,40}$)(?!.*[_.-]{2})[^_.-].*[^_.-]$');
    let passwordRegEx = new RegExp('^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#\$%\^&\*])(?=.{8,})');
    let emailRegEx = new RegExp('^[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,}$');

    let usernameTest = usernameRegEx.test(this.state.username);
    let passwordTest = passwordRegEx.test(this.state.password);
    let emailTest = emailRegEx.test(this.state.email);
    let passwordsMatch = (this.state.password === this.state.confirmPassword);

    var test = this.checkData(usernameTest, passwordTest, emailTest, passwordsMatch);

    /*axios.get('https://evening-oasis-01489.herokuapp.com/')
      .then((response) => {
        console.log(response.data);
      }).catch((error) => {
        console.log(error);
      })
    */

    let salt = bcrypt.genSaltSync(10);
    let hash = bcrypt.hashSync(this.state.password, salt);

    console.log(test + ' test');

    /*if(test) {
      axios.post('https://evening-oasis-01489.herokuapp.com/register', {
        email: this.state.email,
        username: this.state.username,
        password: hash
      })
      .then(function (response) {
        console.log(response.data);
        let verifyID = response.data;
      })
      .catch(function (error) {
        let errorCode = error.response.status;
        if(errorCode == 452) {
          // Email taken
        } else if (errorCode == 453) {
          // Username taken
        } else {
          // Connection error
        }
      });
    }
*/
    //console.log(usernameRegEx.test(this.state.username));
    //console.log(passwordRegEx.test(this.state.password));
    //console.log(emailRegEx.test(this.state.email));
    //console.log(passwordsMatch);

    //this.props.navigation.navigate('Profile');
  }

  render() {

  /*  if (!this.state.isReady) {
      return <AppLoading />;
    }*/

    return (
    <StyleProvider style={getTheme(material)}>
      <Container>
        <Content>
        <AppHeader />
            <Grid>
              <Body>
                <Col style={{ width: Dimensions.get('window').width*0.8 }}>
                  <SignUpAlert errorText={this.state.errorText} invalidData={this.state.invalidData} />
                  <Body>
                    <Row style={{ marginVertical: 10 }}>
                      <H1>
                      Sign up
                      </H1>
                    </Row>
                    <Item rounded>
                      <Input style={{ paddingLeft: 15 }} placeholder='Username' value={this.state.username}
                      onChangeText={user => this.setState({ username: user })}/>
                    </Item>
                    <Item rounded>
                      <Input style={{ paddingLeft: 15 }} placeholder='Email' value={this.state.email}
                      onChangeText={email => this.setState({ email: email })}/>
                    </Item>
                    <Item rounded>
                      <Input style={{ paddingLeft: 15 }} secureTextEntry={true} placeholder='Password' value={this.state.password}
                      onChangeText={password => this.setState({ password: password })}/>
                    </Item>
                    <Item rounded>
                      <Input style={{ paddingLeft: 15 }} secureTextEntry={true} placeholder='Confirm password' value={this.state.confirmPassword}
                      onChangeText={confirm => this.setState({ confirmPassword: confirm })}/>
                    </Item>
                    <Button onPress={() => this.props.navigation.navigate('Login')} transparent>
                      <Text>I already have an account</Text>
                    </Button>
                    <Row style={{ marginVertical: 10 }}>
                      <Button onPress={this.register} style={{ justifyContent: "center", width: 50 }} bordered large rounded>
                        <Icon type='Entypo' name='chevron-right' />
                      </Button>
                    </Row>
                  </Body>
                </Col>
              </Body>
            </Grid>
          </Content>
      </Container>
    </StyleProvider>
    );
  }
}

export default Sign_Up;
