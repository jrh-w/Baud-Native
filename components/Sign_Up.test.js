import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';

import Sign_Up from './Sign_Up';

test('Sign Up component functionality test', async () => {

  const navigation = { // Mock of react-navigation
    navigate: jest.fn()
  };

  const component = render(<Sign_Up navigation={navigation}/>).toJSON();
  const { getAllByA11yLabel, getByPlaceholderText } = render(<Sign_Up navigation={navigation}/>);

  fireEvent.changeText(getByPlaceholderText('Username'), 'asdfghjkl')
  fireEvent.changeText(getByPlaceholderText('Email'), 'jaqub.f.wjlodaz@gmmm.coz')
  fireEvent.changeText(getByPlaceholderText('Password'), 'gendarme21254+')
  fireEvent.changeText(getByPlaceholderText('Confirm password'), 'gendarme21254+')

  const register = fireEvent.press(getAllByA11yLabel('register')[0]);

  expect(register).toBeTruthy();

  expect(component).toMatchSnapshot();
});
