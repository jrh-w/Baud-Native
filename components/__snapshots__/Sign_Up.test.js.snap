// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign Up component functionality test 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F7F7F7",
      "flex": 1,
      "height": 1334,
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "backgroundColor": "transparent",
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      automaticallyAdjustContentInsets={false}
      contentContainerStyle={
        Array [
          Object {
            "padding": undefined,
          },
          undefined,
        ]
      }
      contentInset={
        Object {
          "bottom": 0,
        }
      }
      enableAutomaticScroll={true}
      enableOnAndroid={false}
      enableResetScrollToCoords={true}
      extraHeight={75}
      extraScrollHeight={0}
      getScrollResponder={[Function]}
      handleOnScroll={[Function]}
      keyboardDismissMode="interactive"
      keyboardOpeningTime={250}
      keyboardShouldPersistTaps="handled"
      keyboardSpace={0}
      onScroll={[Function]}
      resetKeyboardSpace={[Function]}
      resetScrollToCoords={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      scrollEventThrottle={1}
      scrollForExtraHeightOnAndroid={[Function]}
      scrollIntoView={[Function]}
      scrollToEnd={[Function]}
      scrollToFocusedInput={[Function]}
      scrollToPosition={[Function]}
      showsVerticalScrollIndicator={true}
      style={
        Object {
          "backgroundColor": "transparent",
          "flex": 1,
        }
      }
      update={[Function]}
      viewIsInsideTabBar={false}
    >
      <View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#fff",
                "borderBottomWidth": 0.5,
                "borderColor": "#ccc",
                "borderLeftWidth": 0.5,
                "borderRadius": 20,
                "borderRightWidth": 0.5,
                "borderTopWidth": 0,
                "elevation": 3,
                "flexWrap": "nowrap",
                "marginBottom": 15,
                "marginTop": 15,
                "shadowColor": "#000",
                "shadowOffset": Object {
                  "height": 3,
                  "width": 0,
                },
                "shadowOpacity": 0.1,
                "shadowRadius": 2.5,
              },
              Object {
                "marginTop": -80,
                "paddingTop": 80,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#fff",
                "borderRadius": 20,
                "flexDirection": "row",
                "paddingBottom": 10,
                "paddingLeft": 15,
                "paddingRight": 15,
                "paddingTop": 10,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "flex-start",
                    "alignSelf": "stretch",
                    "flex": 1,
                  },
                  Object {
                    "alignItems": "center",
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "flex": 1,
                    "flexDirection": "row",
                  }
                }
              >
                <Image
                  source={1}
                  style={
                    Object {
                      "height": 266.8,
                      "width": 266.8,
                    }
                  }
                />
              </View>
              <View
                style={
                  Object {
                    "flex": 1,
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#000",
                        "fontFamily": "Montserrat_Bold",
                        "fontSize": 24,
                        "lineHeight": 32,
                      },
                      Object {
                        "marginBottom": 10,
                        "marginTop": 20,
                      },
                    ]
                  }
                >
                  Baud
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "flex": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 0,
                  "flexDirection": "column",
                  "width": 600,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "alignSelf": "center",
                    "flex": 1,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                      "flexDirection": "row",
                      "marginVertical": 10,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#000",
                        "fontFamily": "Montserrat_Bold",
                        "fontSize": 24,
                        "lineHeight": 32,
                      }
                    }
                  >
                    Sign up
                  </Text>
                </View>
                <View
                  accessible={true}
                  focusable={false}
                  isTVSelectable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#fff",
                      "borderBottomWidth": 1,
                      "borderColor": "#D9D5DC",
                      "borderLeftWidth": 1,
                      "borderRadius": 30,
                      "borderRightWidth": 1,
                      "borderTopWidth": 1,
                      "borderWidth": 1,
                      "flexDirection": "row",
                      "marginBottom": 10,
                      "marginLeft": 2,
                      "marginTop": 10,
                      "opacity": 1,
                      "shadowColor": "#000",
                      "shadowOffset": Object {
                        "height": 3,
                        "width": 0,
                      },
                      "shadowOpacity": 0.1,
                      "shadowRadius": 2.5,
                    }
                  }
                >
                  <TextInput
                    allowFontScaling={true}
                    editable={true}
                    onChangeText={[Function]}
                    placeholder="Username"
                    placeholderTextColor="#575757"
                    rejectResponderTermination={true}
                    style={
                      Array [
                        Object {
                          "color": "#000",
                          "flex": 1,
                          "fontFamily": "SemiBold",
                          "fontSize": 14,
                          "height": 35,
                          "paddingLeft": 8,
                          "paddingRight": 5,
                          "top": 1.5,
                        },
                        Object {
                          "paddingLeft": 15,
                        },
                      ]
                    }
                    underlineColorAndroid="rgba(0,0,0,0)"
                    value=""
                  />
                </View>
                <View
                  accessible={true}
                  focusable={false}
                  isTVSelectable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#fff",
                      "borderBottomWidth": 1,
                      "borderColor": "#D9D5DC",
                      "borderLeftWidth": 1,
                      "borderRadius": 30,
                      "borderRightWidth": 1,
                      "borderTopWidth": 1,
                      "borderWidth": 1,
                      "flexDirection": "row",
                      "marginBottom": 10,
                      "marginLeft": 2,
                      "marginTop": 10,
                      "opacity": 1,
                      "shadowColor": "#000",
                      "shadowOffset": Object {
                        "height": 3,
                        "width": 0,
                      },
                      "shadowOpacity": 0.1,
                      "shadowRadius": 2.5,
                    }
                  }
                >
                  <TextInput
                    allowFontScaling={true}
                    editable={true}
                    onChangeText={[Function]}
                    placeholder="Email"
                    placeholderTextColor="#575757"
                    rejectResponderTermination={true}
                    style={
                      Array [
                        Object {
                          "color": "#000",
                          "flex": 1,
                          "fontFamily": "SemiBold",
                          "fontSize": 14,
                          "height": 35,
                          "paddingLeft": 8,
                          "paddingRight": 5,
                          "top": 1.5,
                        },
                        Object {
                          "paddingLeft": 15,
                        },
                      ]
                    }
                    underlineColorAndroid="rgba(0,0,0,0)"
                    value=""
                  />
                </View>
                <View
                  accessible={true}
                  focusable={false}
                  isTVSelectable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#fff",
                      "borderBottomWidth": 1,
                      "borderColor": "#D9D5DC",
                      "borderLeftWidth": 1,
                      "borderRadius": 30,
                      "borderRightWidth": 1,
                      "borderTopWidth": 1,
                      "borderWidth": 1,
                      "flexDirection": "row",
                      "marginBottom": 10,
                      "marginLeft": 2,
                      "marginTop": 10,
                      "opacity": 1,
                      "shadowColor": "#000",
                      "shadowOffset": Object {
                        "height": 3,
                        "width": 0,
                      },
                      "shadowOpacity": 0.1,
                      "shadowRadius": 2.5,
                    }
                  }
                >
                  <TextInput
                    allowFontScaling={true}
                    editable={true}
                    onChangeText={[Function]}
                    placeholder="Password"
                    placeholderTextColor="#575757"
                    rejectResponderTermination={true}
                    secureTextEntry={true}
                    style={
                      Array [
                        Object {
                          "color": "#000",
                          "flex": 1,
                          "fontFamily": "SemiBold",
                          "fontSize": 14,
                          "height": 35,
                          "paddingLeft": 8,
                          "paddingRight": 5,
                          "top": 1.5,
                        },
                        Object {
                          "paddingLeft": 15,
                        },
                      ]
                    }
                    underlineColorAndroid="rgba(0,0,0,0)"
                    value=""
                  />
                </View>
                <View
                  accessible={true}
                  focusable={false}
                  isTVSelectable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#fff",
                      "borderBottomWidth": 1,
                      "borderColor": "#D9D5DC",
                      "borderLeftWidth": 1,
                      "borderRadius": 30,
                      "borderRightWidth": 1,
                      "borderTopWidth": 1,
                      "borderWidth": 1,
                      "flexDirection": "row",
                      "marginBottom": 10,
                      "marginLeft": 2,
                      "marginTop": 10,
                      "opacity": 1,
                      "shadowColor": "#000",
                      "shadowOffset": Object {
                        "height": 3,
                        "width": 0,
                      },
                      "shadowOpacity": 0.1,
                      "shadowRadius": 2.5,
                    }
                  }
                >
                  <TextInput
                    allowFontScaling={true}
                    editable={true}
                    onChangeText={[Function]}
                    placeholder="Confirm password"
                    placeholderTextColor="#575757"
                    rejectResponderTermination={true}
                    secureTextEntry={true}
                    style={
                      Array [
                        Object {
                          "color": "#000",
                          "flex": 1,
                          "fontFamily": "SemiBold",
                          "fontSize": 14,
                          "height": 35,
                          "paddingLeft": 8,
                          "paddingRight": 5,
                          "top": 1.5,
                        },
                        Object {
                          "paddingLeft": 15,
                        },
                      ]
                    }
                    underlineColorAndroid="rgba(0,0,0,0)"
                    value=""
                  />
                </View>
                <View />
                <View
                  accessible={true}
                  focusable={true}
                  isTVSelectable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "borderBottomWidth": null,
                      "borderColor": "#1DA8F0",
                      "borderLeftWidth": null,
                      "borderRadius": 2,
                      "borderRightWidth": null,
                      "borderTopWidth": null,
                      "borderWidth": undefined,
                      "elevation": 0,
                      "flexDirection": "row",
                      "height": 45,
                      "justifyContent": "space-between",
                      "opacity": 1,
                      "paddingBottom": 6,
                      "paddingTop": 6,
                      "shadowColor": null,
                      "shadowOffset": null,
                      "shadowOpacity": null,
                      "shadowRadius": null,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "backgroundColor": "transparent",
                        "color": "#1DA8F0",
                        "fontFamily": "SemiBold",
                        "fontSize": 14,
                        "marginLeft": 0,
                        "marginRight": 0,
                      }
                    }
                  >
                    I already have an account
                  </Text>
                </View>
                <View
                  style={
                    Object {
                      "flex": 1,
                      "flexDirection": "row",
                      "marginVertical": 10,
                    }
                  }
                >
                  <View
                    accessibilityLabel="register"
                    accessible={true}
                    focusable={true}
                    isTVSelectable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "transparent",
                        "borderBottomWidth": 2,
                        "borderColor": "#1DA8F0",
                        "borderLeftWidth": 2,
                        "borderRadius": 57,
                        "borderRightWidth": 2,
                        "borderTopWidth": 2,
                        "borderWidth": 1,
                        "elevation": null,
                        "flexDirection": "row",
                        "height": 50,
                        "justifyContent": "center",
                        "opacity": 1,
                        "paddingBottom": 6,
                        "paddingTop": 6,
                        "shadowColor": null,
                        "shadowOffset": null,
                        "shadowOpacity": null,
                        "shadowRadius": null,
                        "width": 50,
                      }
                    }
                  >
                    <Text />
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </RCTSafeAreaView>
</View>
`;
